Add Save, SaveAll, Restore, RestoreAll to pkg/util/iptables

BenTheElder opened the pull request:
Adds utility wrappers for `iptables-save` and `iptables-restore` to the
iptables Interface in pkg/util/iptables. Also const’s command strings.

BenTheElder committed:
Add Save, SaveAll, Restore, RestoreAll to pkg/util/iptables

Adds utility wrappers for `iptables-save` and `iptables-restore` to the
iptables Interface in pkg/util/iptables. Also const’s command strings.

BenTheElder commented:
This is part of work for implementing https://github.com/GoogleCloudPlatform/kubernetes/issues/3760; see larger PR: https://github.com/GoogleCloudPlatform/kubernetes/pull/9210
/cc @thockin

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -178,11 +211,112 @@ func (runner *runner) IsIpv6() bool {
+	// create temp file through which to pass data
+	temp, err := ioutil.TempFile("", "kube-temp-iptables-restore-")
+	// make sure we delete the temp file
+	defer os.Remove(temp.Name())
```
only defer if err == nil

BenTheElder replied:
fixed.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -178,11 +211,112 @@ func (runner *runner) IsIpv6() bool {
+	// create temp file through which to pass data
+	temp, err := ioutil.TempFile("", "kube-temp-iptables-restore-")
+	// make sure we delete the temp file
+	defer os.Remove(temp.Name())
```
same comment

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -178,11 +211,112 @@ func (runner *runner) IsIpv6() bool {
+
+	// setup args
+	args := []string{"-T", string(table)}
+	if !flush {
```
can you capture lines 240-271 as a private method and call it from both Restore() and RestoreAll() ?

BenTheElder replied:
yes

thockin replied:
let me try again, I know you CAN, but would you please do so :)

BenTheElder replied:
Understood first time, silly reply on my part just checking off each comment as I fix up the next commit.

thockin replied:
I was being a smart ass - I have a 5 year old who would say yes and stare at me. :)

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -49,6 +51,18 @@ type Interface interface {
 	// IsIpv6 returns true if this is managing ipv6 tables
 	IsIpv6() bool
+	// Save calls `iptables-save` for table.
+	Save(table Table) ([]byte, error)
```
Does the addition of these methods break any other call-sites that assume they are implementing iptables.Interface ?

thockin replied:
argument to createProxier:
*fakeIptables does not implement iptables.Interface (missing Restore method)
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/proxier_test.go:508: cannot use fakeIptables literal (type *fakeIptables) as type iptables.Interface in argument to createProxier:
*fakeIptables does not implement iptables.Interface (missing Restore method)
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/proxier_test.go:560: cannot use fakeIptables literal (type *fakeIptables) as type iptables.Interface in argument to createProxier:
*fakeIptables does not implement iptables.Interface (missing Restore method)
_output/local/go/src/k8s.io/kubernetes/pkg/proxy/proxier_test.go:560: too many errors
FAIL	k8s.io/kubernetes/pkg/proxy [build failed]

BenTheElder replied:
Ah yes, one callsite that mocks it in the userspace proxier tests, will
patch that in shortly. Forgot that, quick-release built fine, tests break
from this.

On Fri, Aug 7, 2015 at 6:09 PM, Tim Hockin <notifications@github.com> wrote:

> In pkg/util/iptables/iptables.go
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12416#discussion_r36565795>
> :
>
> > @@ -49,6 +51,18 @@ type Interface interface {
> >  	DeleteRule(table Table, chain Chain, args ...string) error
> >  	// IsIpv6 returns true if this is managing ipv6 tables
> >  	IsIpv6() bool
> > +	// Save calls `iptables-save` for table.
> > +	Save(table Table) ([]byte, error)
>
> Does the addition of these methods break any other call-sites that assume
> they are implementing iptables.Interface ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12416/files#r36565795>
> .
>

BenTheElder replied:
the code for this exists in the iptables proxy branch, just need to pull over here.

thockin commented on pkg/util/iptables/iptables.go:
```diff
@@ -49,6 +51,18 @@ type Interface interface {
+	// Save calls `iptables-save` for table.
+	Save(table Table) ([]byte, error)
+	// SaveAll calls `iptables-save`.
+	SaveAll() ([]byte, error)
```
These really should have unit tests.  I will accept a TODO as long as you promise to do them :)

The tests won't be hard, but they will ensure that what happens is what we expect.

BenTheElder committed:
Fix iptables Interface mocking, move Restore/RestoreAll to shared impl

also put TODO for unit tests, move defer file deletion until after file
creation error is checked.

thockin commented:
LGTM
